<?xml version="1.0" encoding="utf-8"?>
<!-- =================================================================================== -->
<!-- |    This document was autogenerated by xacro from src/husky_panda/husky_panda_manipulation/data/husky_panda_mobile.urdf.xacro | -->
<!-- |    EDITING THIS FILE BY HAND IS NOT RECOMMENDED                                 | -->
<!-- =================================================================================== -->
<robot name="panda">
  <material name="dark_grey">
    <color rgba="0.2 0.2 0.2 1.0"/>
  </material>
  <material name="medium_grey">
    <color rgba="0.6 0.6 0.6 1.0"/>
  </material>
  <material name="light_grey">
    <color rgba="0.8 0.8 0.8 1.0"/>
  </material>
  <material name="yellow">
    <color rgba="0.8 0.8 0.0 1.0"/>
  </material>
  <material name="black">
    <color rgba="0.15 0.15 0.15 1.0"/>
  </material>
  <material name="white">
    <color rgba="1.0 1.0 1.0 1.0"/>
  </material>
  <material name="red">
    <color rgba="1.0 0.0 0.0 1.0"/>
  </material>
  <!-- Base link is the center of the robot's bottom plate -->
  <link name="base_link">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="../../meshes/husky_panda/husky/base_link.dae"/>
      </geometry>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0.0 0 0.061875"/>
      <geometry>
        <box size="0.9874 0.5709 0.12375"/>
      </geometry>
    </collision>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0.175625"/>
      <geometry>
        <box size="0.78992 0.5709 0.10375"/>
      </geometry>
    </collision>
  </link>
  <!-- Base footprint is on the ground under the robot -->
  <link name="base_footprint"/>
  <joint name="base_footprint_joint" type="fixed">
    <origin rpy="0 0 0" xyz="0 0 -0.13228"/>
    <parent link="base_link"/>
    <child link="base_footprint"/>
  </joint>
  <!-- Inertial link stores the robot's inertial information -->
  <link name="inertial_link">
    <inertial>
      <mass value="46.034"/>
      <origin xyz="-0.00065 -0.085 0.062"/>
      <inertia ixx="0.6022" ixy="-0.02364" ixz="-0.1197" iyy="1.7386" iyz="-0.001544" izz="2.0296"/>
    </inertial>
  </link>
  <joint name="inertial_joint" type="fixed">
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <parent link="base_link"/>
    <child link="inertial_link"/>
  </joint>
  <!-- IMU Link is the standard mounting position for the UM6 IMU.-->
  <!-- Can be modified with environment variables in /etc/ros/setup.bash -->
  <link name="imu_link"/>
  <joint name="imu_joint" type="fixed">
    <origin rpy="0 -1.5708 3.1416" xyz="0.19 0 0.149"/>
    <parent link="base_link"/>
    <child link="imu_link"/>
  </joint>
  <gazebo reference="imu_link">
  </gazebo>
  <link name="front_left_wheel_link">
    <inertial>
      <mass value="2.637"/>
      <origin xyz="0 0 0"/>
      <inertia ixx="0.02467" ixy="0" ixz="0" iyy="0.04411" iyz="0" izz="0.02467"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="../../meshes/husky_panda/husky/wheel.dae"/>
      </geometry>
    </visual>
    <collision>
      <origin rpy="1.570795 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.1143" radius="0.1651"/>
      </geometry>
    </collision>
    <contact>
      <lateral_friction value="1.0"/>
      <rolling_friction value="0.0"/>
      <stiffness value="30000"/>
      <damping value="1000"/>
    </contact>
  </link>
  <gazebo reference="front_left_wheel_link">
    <mu1 value="1.0"/>
    <mu2 value="1.0"/>
    <kp value="10000000.0"/>
    <kd value="1.0"/>
    <fdir1 value="1 0 0"/>
  </gazebo>
  <joint name="front_left_wheel" type="continuous">
    <parent link="base_link"/>
    <child link="front_left_wheel_link"/>
    <origin rpy="0 0 0" xyz="0.256 0.2854 0.03282"/>
    <axis rpy="0 0 0" xyz="0 1 0"/>
  </joint>
  <!-- <joint name="${wheel_prefix}_wheel" type="continuous">
			<parent link="base_link"/>
			<child link="${wheel_prefix}_wheel_link"/>
			<xacro:insert_block name="joint_pose"/>
			<axis xyz="0 1 0" rpy="0 0 0" />
		</joint>

		<transmission name="${wheel_prefix}_wheel_trans" type="SimpleTransmission">
			<type>transmission_interface/SimpleTransmission</type>
			<actuator name="${wheel_prefix}_wheel_motor">
				<mechanicalReduction>1</mechanicalReduction>
			</actuator>
			<joint name="${wheel_prefix}_wheel">
				<hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
			</joint>
		</transmission> -->
  <link name="front_right_wheel_link">
    <inertial>
      <mass value="2.637"/>
      <origin xyz="0 0 0"/>
      <inertia ixx="0.02467" ixy="0" ixz="0" iyy="0.04411" iyz="0" izz="0.02467"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="../../meshes/husky_panda/husky/wheel.dae"/>
      </geometry>
    </visual>
    <collision>
      <origin rpy="1.570795 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.1143" radius="0.1651"/>
      </geometry>
    </collision>
    <contact>
      <lateral_friction value="1.0"/>
      <rolling_friction value="0.0"/>
      <stiffness value="30000"/>
      <damping value="1000"/>
    </contact>
  </link>
  <gazebo reference="front_right_wheel_link">
    <mu1 value="1.0"/>
    <mu2 value="1.0"/>
    <kp value="10000000.0"/>
    <kd value="1.0"/>
    <fdir1 value="1 0 0"/>
  </gazebo>
  <joint name="front_right_wheel" type="continuous">
    <parent link="base_link"/>
    <child link="front_right_wheel_link"/>
    <origin rpy="0 0 0" xyz="0.256 -0.2854 0.03282"/>
    <axis rpy="0 0 0" xyz="0 1 0"/>
  </joint>
  <!-- <joint name="${wheel_prefix}_wheel" type="continuous">
			<parent link="base_link"/>
			<child link="${wheel_prefix}_wheel_link"/>
			<xacro:insert_block name="joint_pose"/>
			<axis xyz="0 1 0" rpy="0 0 0" />
		</joint>

		<transmission name="${wheel_prefix}_wheel_trans" type="SimpleTransmission">
			<type>transmission_interface/SimpleTransmission</type>
			<actuator name="${wheel_prefix}_wheel_motor">
				<mechanicalReduction>1</mechanicalReduction>
			</actuator>
			<joint name="${wheel_prefix}_wheel">
				<hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
			</joint>
		</transmission> -->
  <link name="rear_left_wheel_link">
    <inertial>
      <mass value="2.637"/>
      <origin xyz="0 0 0"/>
      <inertia ixx="0.02467" ixy="0" ixz="0" iyy="0.04411" iyz="0" izz="0.02467"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="../../meshes/husky_panda/husky/wheel.dae"/>
      </geometry>
    </visual>
    <collision>
      <origin rpy="1.570795 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.1143" radius="0.1651"/>
      </geometry>
    </collision>
    <contact>
      <lateral_friction value="1.0"/>
      <rolling_friction value="0.0"/>
      <stiffness value="30000"/>
      <damping value="1000"/>
    </contact>
  </link>
  <gazebo reference="rear_left_wheel_link">
    <mu1 value="1.0"/>
    <mu2 value="1.0"/>
    <kp value="10000000.0"/>
    <kd value="1.0"/>
    <fdir1 value="1 0 0"/>
  </gazebo>
  <joint name="rear_left_wheel" type="continuous">
    <parent link="base_link"/>
    <child link="rear_left_wheel_link"/>
    <origin rpy="0 0 0" xyz="-0.256 0.2854 0.03282"/>
    <axis rpy="0 0 0" xyz="0 1 0"/>
  </joint>
  <!-- <joint name="${wheel_prefix}_wheel" type="continuous">
			<parent link="base_link"/>
			<child link="${wheel_prefix}_wheel_link"/>
			<xacro:insert_block name="joint_pose"/>
			<axis xyz="0 1 0" rpy="0 0 0" />
		</joint>

		<transmission name="${wheel_prefix}_wheel_trans" type="SimpleTransmission">
			<type>transmission_interface/SimpleTransmission</type>
			<actuator name="${wheel_prefix}_wheel_motor">
				<mechanicalReduction>1</mechanicalReduction>
			</actuator>
			<joint name="${wheel_prefix}_wheel">
				<hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
			</joint>
		</transmission> -->
  <link name="rear_right_wheel_link">
    <inertial>
      <mass value="2.637"/>
      <origin xyz="0 0 0"/>
      <inertia ixx="0.02467" ixy="0" ixz="0" iyy="0.04411" iyz="0" izz="0.02467"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="../../meshes/husky_panda/husky/wheel.dae"/>
      </geometry>
    </visual>
    <collision>
      <origin rpy="1.570795 0 0" xyz="0 0 0"/>
      <geometry>
        <cylinder length="0.1143" radius="0.1651"/>
      </geometry>
    </collision>
    <contact>
      <lateral_friction value="1.0"/>
      <rolling_friction value="0.0"/>
      <stiffness value="30000"/>
      <damping value="1000"/>
    </contact>
  </link>
  <gazebo reference="rear_right_wheel_link">
    <mu1 value="1.0"/>
    <mu2 value="1.0"/>
    <kp value="10000000.0"/>
    <kd value="1.0"/>
    <fdir1 value="1 0 0"/>
  </gazebo>
  <joint name="rear_right_wheel" type="continuous">
    <parent link="base_link"/>
    <child link="rear_right_wheel_link"/>
    <origin rpy="0 0 0" xyz="-0.256 -0.2854 0.03282"/>
    <axis rpy="0 0 0" xyz="0 1 0"/>
  </joint>
  <!-- <joint name="${wheel_prefix}_wheel" type="continuous">
			<parent link="base_link"/>
			<child link="${wheel_prefix}_wheel_link"/>
			<xacro:insert_block name="joint_pose"/>
			<axis xyz="0 1 0" rpy="0 0 0" />
		</joint>

		<transmission name="${wheel_prefix}_wheel_trans" type="SimpleTransmission">
			<type>transmission_interface/SimpleTransmission</type>
			<actuator name="${wheel_prefix}_wheel_motor">
				<mechanicalReduction>1</mechanicalReduction>
			</actuator>
			<joint name="${wheel_prefix}_wheel">
				<hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
			</joint>
		</transmission> -->
  <!-- Spawn Husky chassis -->
  <link name="top_chassis_link">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="../../meshes/husky_panda/husky/top_chassis.dae"/>
      </geometry>
    </visual>
  </link>
  <!-- Attach chassis to the robot -->
  <joint name="top_chassis_joint" type="fixed">
    <parent link="base_link"/>
    <child link="top_chassis_link"/>
  </joint>
  <!-- Spawn user rails -->
  <link name="user_rail_link">
    <visual>
      <geometry>
        <mesh filename="../../meshes/husky_panda/husky/user_rail.dae"/>
      </geometry>
    </visual>
  </link>
  <!-- Attach user rails to base link -->
  <joint name="user_rail" type="fixed">
    <origin rpy="0 0 0" xyz="0.272 0 0.245"/>
    <parent link="base_link"/>
    <child link="user_rail_link"/>
  </joint>
  <link name="front_bumper_link">
    <visual>
      <geometry>
        <mesh filename="../../meshes/husky_panda/husky/bumper.dae"/>
      </geometry>
    </visual>
  </link>
  <!-- Attach front bumper -->
  <joint name="front_bumper" type="fixed">
    <origin rpy="0 0 0" xyz="0.48 0 0.091"/>
    <parent link="base_link"/>
    <child link="front_bumper_link"/>
  </joint>
  <link name="rear_bumper_link">
    <visual>
      <geometry>
        <mesh filename="../../meshes/husky_panda/husky/bumper.dae"/>
      </geometry>
    </visual>
  </link>
  <!-- Attach rear bumper -->
  <joint name="rear_bumper" type="fixed">
    <origin rpy="0 0 3.14159" xyz="-0.48 0 0.091"/>
    <parent link="base_link"/>
    <child link="rear_bumper_link"/>
  </joint>
  <link name="top_plate_link">
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="../../meshes/husky_panda/husky/top_plate.dae"/>
      </geometry>
    </visual>
    <!-- <collision>
            <origin xyz="0 0 0" rpy="0 0 0"/>
            <geometry>
              <mesh filename="../../meshes/husky_panda/husky/top_plate.dae" />
            </geometry>
          </collision> -->
  </link>
  <!-- Attach top plate -->
  <joint name="top_plate_joint" type="fixed">
    <parent link="base_link"/>
    <child link="top_plate_link"/>
    <origin rpy="0 0 0" xyz="0.0812 0 0.245"/>
  </joint>
  <!-- Top plate front link -->
  <joint name="top_plate_front_joint" type="fixed">
    <parent link="top_plate_link"/>
    <child link="top_plate_front_link"/>
    <origin rpy="0 0 0" xyz="0.36367 0 0.00639"/>
  </joint>
  <!-- Top plate rear link-->
  <joint name="top_plate_rear_joint" type="fixed">
    <parent link="top_plate_link"/>
    <child link="top_plate_rear_link"/>
    <origin rpy="0 0 0" xyz="-0.36633 0 0.00639"/>
  </joint>
  <link name="top_plate_front_link"/>
  <link name="top_plate_rear_link"/>

  <joint name="front_laser_to_top_plate_front_link" type="fixed">
    <parent link="top_plate_front_link"/>
    <child link="front_laser_mount_link"/>
    <origin rpy="0 0 0.785398163" xyz="0 0.25 0"/>
  </joint>
  <link name="front_laser_mount_link">
    <inertial>
        <mass value="0.001"/>
        <origin xyz="0 0 0"/>
        <inertia ixx="0.000001" ixy="0.0" ixz="0.0" iyy="0.000001" iyz="0.0" izz="0.000001"/>
    </inertial>
  </link>
  <joint name="front_laser_joint" type="fixed">
    <parent link="front_laser_mount_link"/>
    <child link="front_laser"/>
    <origin rpy="0 0 0" xyz="0 0 0.05595"/>
  </joint>
  <link name="front_laser">
    <visual>
      <geometry>
        <mesh filename="../../meshes/husky_panda/husky/accessories/sick_tim_5xx.stl"/>
      </geometry>
      <material name="blue">
        <color rgba="0 0 1 1"/>
      </material>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <box size="0.05 0.05 0.05"/>
      </geometry>
      <!-- <geometry>
        <mesh filename="../../meshes/husky_panda/husky/accessories/sick_tim_5xx.stl"/>
      </geometry> -->
    </collision>
    <inertial>
      <mass value="0.15"/>
      <origin xyz="0 0 -0.026"/>
      <inertia ixx="0.000123012450795" ixy="0.0" ixz="0.0" iyy="0.000123012450795" iyz="0.0" izz="8.9999964e-05"/>
    </inertial>
  </link>
  
  <joint name="rear_laser_to_top_plate_rear_link" type="fixed">
    <parent link="top_plate_rear_link"/>
    <child link="rear_laser_mount_link"/>
    <origin rpy="0 0 -2.35619449" xyz="0 -0.25 0"/>
  </joint>
  <link name="rear_laser_mount_link">
    <inertial>
        <mass value="0.001"/>
        <origin xyz="0 0 0"/>
        <inertia ixx="0.000001" ixy="0.0" ixz="0.0" iyy="0.000001" iyz="0.0" izz="0.000001"/>
    </inertial>
  </link>
  <joint name="rear_laser_joint" type="fixed">
    <parent link="rear_laser_mount_link"/>
    <child link="rear_laser"/>
    <origin rpy="0 0 0" xyz="0 0 0.05595"/>
  </joint>
  <link name="rear_laser">
    <visual>
      <geometry>
        <mesh filename="../../meshes/husky_panda/husky/accessories/sick_tim_5xx.stl"/>
      </geometry>
      <material name="blue">
        <color rgba="0 0 1 1"/>
      </material>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <box size="0.05 0.05 0.05"/>
      </geometry>
      <!-- <geometry>
        <mesh filename="../../meshes/husky_panda/husky/accessories/sick_tim_5xx.stl"/>
      </geometry> -->
    </collision>
    <inertial>
      <mass value="0.15"/>
      <origin xyz="0 0 -0.026"/>
      <inertia ixx="0.000123012450795" ixy="0.0" ixz="0.0" iyy="0.000123012450795" iyz="0.0" izz="8.9999964e-05"/>
    </inertial>
  </link>
</robot>